### DISCUSSION TITLE
Let's flash the Shelly Pro 3

### DISCUSSION THREAD

**Discussion by DerT94** [arendst/Tasmota#20230] [General]
Let's flash the Shelly Pro 3 with Tasmota

## ToDo's

- [x] Get a serial conection
- [x] Backup the original firmware
- [x] Flash Tasmota
- [x] Find out GPIO configuration and test functions
- [x] Create template
- [x] Adjustments to Tasmota (if necessary)

I have the device here and have already completely disassembled it. 

### Shelly_Pro3_Main_V1.2.3
<img src="https://github.com/arendst/Tasmota/assets/58786605/42f3ea97-04f9-4d42-b9b2-8d8c9e803980" width="250" height="250">
<img src="https://github.com/arendst/Tasmota/assets/58786605/ff1f17fc-4a05-47f3-b28a-2245610bfb96" width="250" height="250">
<img src="https://github.com/arendst/Tasmota/assets/58786605/958af27e-97f6-4e13-b328-63999b2ea4e4" width="250" height="250">
<img src="https://github.com/arendst/Tasmota/assets/58786605/a3f2814c-0a59-48c1-b7d0-8a4fbe703b06" width="250" height="250">
<img src="https://github.com/arendst/Tasmota/assets/58786605/86005ca5-50fb-4154-ac66-33e7e8d43575" width="250" height="250">

### Shelly_Pro3_Reilays_V2.2.3
<img src="https://github.com/arendst/Tasmota/assets/58786605/815d977a-73c4-4088-bfc6-2f5783df4eca" width="250" height="250">
<img src="https://github.com/arendst/Tasmota/assets/58786605/fa0f8351-1e13-47d1-8bd5-e5e9686dbdee" width="250" height="250">

### Shelly_Pro3_Buttons_V1.2.3
<img src="https://github.com/arendst/Tasmota/assets/58786605/5eeec2c4-122d-4447-bc3a-8fc939177ffe" width="250" height="250">
<img src="https://github.com/arendst/Tasmota/assets/58786605/56f36edc-beae-4c30-98f0-a50a09764037" width="250" height="250">

### Get a Serial Connection
There aresmall 6-pin connector on the back of the "Shelly_Pro3_Buttons_V1.2.3" PCB. These are the usual GPIOs.
<img src="https://github.com/arendst/Tasmota/assets/58786605/352f1e8a-27a6-4582-bd76-abc74ec5741b" width="250" height="250">

### Chip Details
<details><summary>Details</summary>
<p>

- Chip is ESP32-D0WDQ6 (revision v1.0)
- Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
- Crystal is 40MHz
- Manufacturer: 1c
- Device: 7017
- Detected flash size: 8MB

</p>
</details> 

### Backup the original firmware
Backup created
### Flash Tasmota
Flashed [tasmota32.factory.bin](https://ota.tasmota.com/tasmota32/tasmota32.factory.bin)

### Find out GPIO configuration and test functions

- [x] Relay GPIOs
Relay 1 - GPIO 2
Relay 2 - GPIO 4
Relay 3 - GPIO 32

- [x] Switch GPIOs
Switch 1 - 34
Switch 2 - 37
Switch 3 - 38

- [x] Ethernet
GPIO18 = ETH MDIO
GPIO23 = ETH MDC

- [ ] Bluetooth

- [x] Temperature Sensor
GPIO36  = ADC Temp 1

To correct the temperature values
```
AdcParam1 2,10000,10000,3350
```

### Create template

Current template (Work in progress)
```
{"NAME":"ShellyPro3","GPIO":[1,1,224,1,225,1,1,1,1,1,1,1,1,1,5600,1,0,1,1,5568,0,1,1,1,0,0,0,0,226,1,96,1,4736,97,162,1],"FLAG":0,"BASE":1}
```

### Adjustments to Tasmota (if necessary)
Not yet started

**Comment by sfromis** [comment]
Did you follow the usual step of grounding gpio0 before/while applying power, to get the MCU into flashing mode=

**Reply by DerT94** [reply]
Yes
Tried it with gpio0 to GND on Boot and directly after the Boot. 
no success

**Comment by andreas-bulling** [comment]
Could you try tracing the pins back to the ESP - or, the other way around, tracing the correct ESP pins required for flashing to where they go on the board?

**Reply by DerT94** [reply]
Done.

I used this [datasheet](https://www.espressif.com/sites/default/files/documentation/esp32_datasheet_en.pdf).
<img src="https://github.com/arendst/Tasmota/assets/58786605/ca936f8e-a874-419e-9e95-aa673d60425c" width="250" height="250">

It is exactly as expected.
<img src="https://github.com/arendst/Tasmota/assets/58786605/3cb48c5f-4e45-43e0-85c0-bec3cf77db3a" width="250" height="250">

Either something is faulty or I'm missing something.
I have just connected a Shelly 1 to my FTDI adapter.
I get data from it. So the adapter is working.

**Reply by sfromis** [reply]
It is **not** safe to assume that the USB/serial adapter is working for all devices just because you had success with one. Current needs might be different. And many adapters are known to be trouble-makers. You do need to make sure to have one able to provide **at least** 150 mA of 3.3V.

**Reply by DerT94** [reply]
Got it
Apparently my Shelly was faulty. I tested a new one and now get a connection

**Comment by DerT94** [comment]
To be on the safe side
esptool detects a flash size of 8MB. The datasheet of the chip says the same

So the correct command to backup the firmware is
.\esptool.exe --port COM4 read_flash **0x00000 0x800000** Backup_ShellyPro3.bin

After the backup, I would use the "tasmota32.bin" firmware and follow the instructions for new devices.

**Comment by DerT94** [comment]
I flashed Tasmota and have a few Questions.
Tasmota: [tasmota32.factory.bin](https://ota.tasmota.com/tasmota32/tasmota32.factory.bin)

### Reset
For reasons for which I have not yet found a cause, the Shelly keeps resetting itself

Console Output:
``` 
22:31:11.738 RSL: RESULT = {"POWER2":"OFF"}
22:31:11.739 RSL: POWER2 = OFF
22:31:13.920 RSL: RESULT = {"POWER3":"ON"}
22:31:13.922 RSL: POWER3 = ON
22:31:14.994 RSL: RESULT = {"POWER3":"OFF"}
22:31:14.996 RSL: POWER3 = OFF
22:31:21.441 RSL: RESULT = {"Reset":"Reset and Restarting"}
22:31:21.545 CFG: Use defaults
```

### Temparature Sensors
The other Shelly Pro Models use GPIO36 for that. But wich Option I have to configure?

**Reply by barbudor** [reply]
Does the reset and restart occurs after 40sec?
Typical of Button 1 badly configured and triggering self recovery
If the GPIO for Button 1 is correct, you may need the inverted mode

Did you followed the protocol I wrote in the other thread?

**Reply by sfromis** [reply]
That looks a lot like a button input pin registering as "pushed" for 40 seconds, which triggers a Tasmota recovery feature to reset the device. If the Button is on the correct gpio and properly working, you may have to switch from Button to Button_i (or vice versa).

**Reply by DerT94** [reply]
Thank's
The buttons have to be configured in inverted mode.
@barbudor
Yes I followed your guide. But I haven't tested the network yet.

I have now configured one input as Button_I and one as a switch.
The switch switches the relay much faster than the button
Switch: Instant
Button: 1-2 seconds
Is this normal or an error in my configuration

Edit: SetOption13 does the trick.

**Comment by DerT94** [comment]
I tested most of the features, and they all seem to be working well.

I updated the initial post in this discussion with the current progress.
How can I test the Bluetooth function?
 
Although I don't plan to use it, I want to ensure that the Pro 3 can be used with Tasmota to its fullest potential.

Do any other essential functions require testing?

My temperature is approximately 10 degrees higher than the room temperature. I believe the temperature inside the shell is warmer.

@arendst 
What else needs to be done so that you can add the template to Tasmota?

**Reply by barbudor** [reply]
> How can I test the Bluetooth function?

You should be able to load tasmota32-bluetooth on it

Regarding the template, you can with nifty nicy pretty button on the bottom left of the template site
![image](https://github.com/arendst/Tasmota/assets/2996042/96bf5b8c-ef34-42e9-9b30-22962a83f4ea)

That will create a basic template page.
You may be able to link to that discussion.
If not, once the basic template has been created from the form, you can improve it with a PR to https://github.com/blakadder/templates/
This how you can copy the pictures or make a long descriptin
But linkink to the discussion here is good enough as everything is here and you alerady summarize the status at the top of the page 👍

**Reply by sfromis** [reply]
When the functions, including energy monitoring works ok, next step is simply to submit a template. As bluetooth usage is outside of the template, this is not a key point to test. https://templates.blakadder.com/ has a "Add New Template" link for you :smile: 
Also a nice place to include comments like special considerations or things to remember, if there is something "less intuitive".

**Comment by Mr-Mime** [comment]
Hey @DerT94 whats the state of the tasmota integration?

From the checklist and the comments it seems like you got the Pro 3 working, whats your experience?

Is only the "official" support missing, aka creating and adding a new template?!

Thanks in advance! :)

**Reply by DerT94** [reply]
You can use the Template from my initial Post. I tested some Szenarios and found no Problems. At this moment I don‘t use the pro 3 productive, only in test Szenarios.

but with the Template all relays and Inputs are working,

**Reply by barbudor** [reply]
@DerT94 If this is complete, could you please add it to the template repository  https://templates.blakadder.com/index.html
Button on the bottom left "ADD NEW TEMPLATE"
Thanks

**Reply by Mr-Mime** [reply]
Thanks for the fast reply.

I'll order a Pro 3 and will follow the provides steps. :+1:

**Reply by DerT94** [reply]
@barbudor 
I had actually already done that

**Comment by Mr-Mime** [comment]
Device arrived yesterday and the basic flashing worked easily, thanks @DerT94 for figuring out the pins etc.


### Switch Input Config
While the relay logic works from the webserver, the switches S1 and S2 of the Pro 3 do not work as expected.
As it can be seen on the drawing of the case, those are expected to be switches as input.
The problem is, that S1 and S2 are configured as buttons and not switches.
After changing that, it they worked like a charm.
![image](https://github.com/arendst/Tasmota/assets/37423773/aca16c1d-f012-48b8-8114-8170129e724f)


### Reset Button
Another thing was, that on the first try I entered wrong wifi creds and thus was not able to access the Pro 3 anymore.
Was not possible to activate the local access-point.
I saw that the reset button is not assigned to a function.
So I followed the traces and saw it being connected to GPIO35, which I made the `Button 1` in the Template.
Thus it has now the following behavior of the [multi press functions](https://tasmota.github.io/docs/Buttons-and-Switches/#multi-press-functions):

| Input | Action |
| --- | --- |
|1x short press | Toggle relay 1 |
|2x short press | Toggle relay 2 |
|3x short press | Toggle relay 3 |
|6x short press | Start WiFiConfig |

### WiFi LED
The wifi LED is a RGB LED.
I turned off the Blue LED, Green is always on and Red is the `LEDLinki`
This way it will flash Orange when not connected to the Wifi or MQTT and light green, when everything is alright.


### ADC Calibration
Also added the ADC calibration to the template.


### Updated template
```
{"NAME":"ShellyPro3","GPIO":[1,1,224,1,225,1,1,1,3872,544,3840,1,1,1,5600,1,0,1,1,5568,0,1,1,1,0,0,0,0,226,1,160,32,4736,161,162,1],"FLAG":0,"BASE":1,"CMND":"AdcParam1 2,10000,10000,3350"}
```

**Reply by ms1958** [reply]
@Mr-Mime , 43 / 5.000
How did you turn off the blue LED?

**Comment by felixkroeger** [comment]
After installing the Firmware mentioned above with esptool, the Shelly continouisly resets itself. High power consumption should not be aproblem, because this also happens if the Shelly is connected to 230 VAC. 
This is the error log:

00:00:00.002 HDW: ESP32-D0WDQ6 v1.1
00:00:00.020 UFS: FlashFS mounted with 4408 kB free
00:00:00.027 CFG: Use defaults
00:00:00.142 FRC: Some settings have been reset (2)
00:00:01.359 BRY: Berry initialized, RAM used 4978 bytes
00:00:01.373 Project tasmota - Tasmota Version 15.5.0.2(2f136a3-tasmota32)-3.3.8(2026-07-23T20:12:55)
00:00:01.834 WIF: WifiManager active for 3 minutes
ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:288
load:0x40078000,len:13400
load:0x40080400,len:2904
entry 0x40080554

To you have any suggestions?

**Reply by sfromis** [reply]
Assuming that you mean a quick boot loop (instead of 3 minutes intervals), one possible cause is that your power source (like whatever USB/serial adapter you use) is not up to supplying sufficient power for the ESP32 to become active on Wifi. Yes, you said "it" also happens with mains power, but not clear if you actually followed serial also here, which for safety reasons is not to be recommended with AC power.
